src/Sylius/Bundle/PaymentBundle/Form/Type/PaymentMethodChoiceType.php

```diff
@@ -22,39 +27,97 @@
  * @author Paweł Jędrzejewski <pawel@sylius.org>
  * @author Arnaud Langlade <arn0d.dev@gmail.com>
  * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ * @author Anna Walasek <anna.walasek@lakion.com>
  */
-class PaymentMethodChoiceType extends ResourceChoiceType
+class PaymentMethodChoiceType extends AbstractType
 {
     /**
+     * @var MethodsResolverInterface
+     */
+    protected $paymentMethodsResolver;
+
+    /**
+     * @var RepositoryInterface
+     */
+    protected $paymentMethodRepository;
+
+    /**
+     * @param MethodsResolverInterface $paymentMethodsResolver
+     * @param RepositoryInterface $paymentMethodRepository
+     */
+    public function __construct(
+        MethodsResolverInterface $paymentMethodsResolver,
+        RepositoryInterface $paymentMethodRepository
+    ) {
+        $this->paymentMethodsResolver = $paymentMethodsResolver;
+        $this->paymentMethodRepository = $paymentMethodRepository;
+    }
+
+    /**
      * {@inheritdoc}
      */
-    public function configureOptions(OptionsResolver $resolver)
+    public function buildForm(FormBuilderInterface $builder, array $options)
     {
-        parent::configureOptions($resolver);
+        if ($options['multiple']) {
+            $builder->addModelTransformer(new CollectionToArrayTransformer());
+        }
+    }
 
-        $queryBuilder = function (Options $options) {
-            $repositoryOptions = [
-                'disabled' => $options['disabled'],
-            ];
+    /**
+     * {@inheritdoc}
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $choiceList = function (Options $options) {
+            if (isset($options['subject'])) {
+                $resolvedMethods = $this->paymentMethodsResolver->getSupportedMethods($options['subject']);
+            } else {
+                $resolvedMethods = $this->paymentMethodRepository->findAll();
+            }
 
-            return function (PaymentMethodRepositoryInterface $repository) use ($repositoryOptions) {
-                return $repository->getQueryBuilderForChoiceType($repositoryOptions);
-            };
+            return new ObjectChoiceList($resolvedMethods, null, [], null, 'id');
         };
 
         $resolver
             ->setDefaults([
-                'query_builder' => $queryBuilder,
-                'disabled' => false,
+                'choice_list' => $choiceList,
             ])
+            ->setDefined([
+                'subject',
+            ])
+            ->setAllowedTypes('subject', PaymentInterface::class)
         ;
     }
 
     /**
+     * {@inheritdoc}
+     */
+    public function getParent()
+    {
+        return 'choice';
+    }
+
+    /**
      * @return string
      */
     public function getName()
     {
         return 'sylius_payment_method_choice';
     }
+
+    /**
+     * @param Options $options
+     * 
+     * @return ObjectChoiceList
+     */
+    private function createChoiceList(Options $options)
```

Comment:
I cannot decide. Should it be extracted? In other class it is always anonymous fuction.